kratos/sources/model_part.cpp

```diff
@@ -1891,12 +1891,29 @@ void ModelPart::AddGeometry(
     typename GeometryType::Pointer pNewGeometry)
 {
     if (IsSubModelPart()) {
-        if (!mpParentModelPart->HasGeometry(pNewGeometry->Id())) {
-            mpParentModelPart->AddGeometry(pNewGeometry);
+        mpParentModelPart->AddGeometry(pNewGeometry);
+        mGeometries.insert(pNewGeometry);
+    }
+    else /// Check if geometry id already used, is done within the geometry container.
+    {
+        auto existing_geometry_it = mGeometries.find(pNewGeometry->Id());
+        if( existing_geometry_it == mGeometries.end()) //node did not exist
+        {
+            mGeometries.insert(pNewGeometry);
+        }
+        else 
+        { // Check if the connectivities coincide
+            // First check for the geometry type
+            KRATOS_ERROR_IF_NOT(GeometryType::HasSameGeometryType(*existing_geometry_it, *pNewGeometry)) << "Attempting to add geometry with Id: " << pNewGeometry->Id() << ". A different geometry with the same Id already exists." << std::endl;
+            // Check that the connectivities are the same
+            // note that we deliberately check the node ids and not the pointer adresses as there might be very rare situations
+
+            // (e.g., creating nodes bypassing the model part interface) with same connectivities but different pointer addresses
```

Comment:
I didn't understand why pointer would fail. Nevertheless, checking the Id is ok for me.

Reply:
I think they're mimicking what we did for elements/conditions not many years ago (I remember discussing this in the @KratosMultiphysics/technical-committee). Am I wrong @rickyaristio?

Reply:
@rubenzorrilla , you are right.